src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs

```diff
@@ -37,17 +37,17 @@ internal static bool TryGetCount<T>(this IEnumerable<T> sequence, out int count)
         /// <returns><c>true</c> if the count could be determined; <c>false</c> otherwise.</returns>
         internal static bool TryGetCount<T>(this IEnumerable sequence, out int count)
         {
-            var collection = sequence as ICollection;
-            if (collection != null)
+            var collectionOfT = sequence as ICollection<T>;
+            if (collectionOfT != null)
             {
-                count = collection.Count;
+                count = collectionOfT.Count;
                 return true;
             }
```

Comment:
IIRC, If you have an array, accessing it's implementation if `ICollection` is faster than accessing its' `ICollection<T>`, and faster still than access its `ICollection<T>` if it is implementing it due to array variance (it isn't actually a a `T[]` but something assignable to `T[]`). Therefore unless you are going to make use of another method or property of `ICollection<T>` at this point (like it's typed `CopyTo`), then the order that was there is probably better in that case.

If something implemented `ICollection<T>` and also `ICollection` it makes no odds.

If something only implements `ICollection<T>` it's probably still faster to find out that it doesn't, than to get `ICollection<T>` for an array.

So that's the details I can think of affecting the balance. Lacking data on actual performance, I'd lean toward having it as it was.